File: Editor/src/main/java/org/turbomedia/turboedit/editor/events/EventSystem.java
package org.turbomedia.turboedit.editor.events;

import java.util.HashMap;
import java.util.UUID;

public class EventSystem {

    private static final HashMap<String, Event> listeners = new HashMap<>();

    public static String RegisterListener(EventType type, EventAction function) {
        var id = UUID.randomUUID().toString();

        listeners.put(id, new Event(type, function));

        return id;
    }

    public static void RemoveListener(String id) {
        listeners.remove(id);
    }

    public static void CallEvent(EventType type, Object data) {
        var map = (HashMap<String, Event>) listeners.clone();

        for (var event : map.values()) {
            if (event.type == type) {
                event.action.onEvent(data);
            }
        }
    }

    public record Event(EventType type, EventAction action) {
    }

    public interface EventAction {
        void onEvent(Object data);
    }
}

File: Editor/src/main/kotlin/turboedit/editor/events/EventSystem.kt
package turboedit.editor.events

import java.util.*

object EventSystem {
    private val listeners = HashMap<String?, Event?>()

    @JvmStatic
    fun registerListener(type: EventType, function: EventAction): String {
        val id: String = UUID.randomUUID().toString()

        listeners.put(id, Event(type, function))

        return id
    }

    fun removeListener(id: String?) {
        listeners.remove(id)
    }

    @JvmStatic
    fun callEvent(type: EventType, data: Any? = null) {
        val map = listeners.clone() as HashMap<String, Event>

        for (event in map.values) {
            if (event.type == type) {
                event.action.onEvent(data)
            }
        }
    }

    @JvmRecord
    data class Event(val type: EventType, val action: EventAction)

    fun interface EventAction {
        fun onEvent(data: Any?)
    }
}
